lib/std/fs.zig

```diff
@@ -851,6 +811,52 @@ pub const Dir = struct {
         }
     }
 
+    pub fn makeDir(self: Dir, sub_path: []const u8) !void {
+        try os.mkdirat(self.fd, sub_path, default_new_dir_mode);
+    }
+
+    pub fn makeDirC(self: Dir, sub_path: [*:0]const u8) !void {
+        try os.mkdiratC(self.fd, sub_path, default_new_dir_mode);
+    }
+
+    /// Calls makeDir recursively to make an entire path. Returns success if the path
+    /// already exists and is a directory.
+    /// This function is not atomic, and if it returns an error, the file system may
+    /// have been modified regardless.
+    pub fn makePath(self: Dir, sub_path: []const u8) !void {
+        var end_index: usize = sub_path.len;
+        while (true) {
+            self.makeDir(sub_path[0..end_index]) catch |err| switch (err) {
```

Comment:
Wow, looks like windows supports this with NtCreateFile

Reply:
> It's too bad we don't have a syscall to "atomically make directory and open the dirfd". E.g. `openat` with `O_CREAT|O_DIRECTORY`. This would allow us to implement `makePath` a bit less racey.

Funny you mention that; I was talking about that just this week :)

 

> looks like windows supports this with NtCreateFile

Yep.